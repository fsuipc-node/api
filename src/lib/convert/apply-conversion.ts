import { VM } from 'vm2';

import { Offset } from '@shared/offset';
import { RawOffsetValue } from '@shared/offset-values';
import { ConvertedOffsetValue } from '@shared/converted-offset-values';

export const applyConversion = (offset: Offset, rawOffsetValue: RawOffsetValue): ConvertedOffsetValue => {
  if (!offset.convert) {
    return rawOffsetValue;
  }

  if (offset.hasMapping) {
    if (offset.isMappingInvalid) {
      return 'INVALID_MAPPING_FUNCTION';
    }

    return offset.mappingFunction(rawOffsetValue);
  } else {
    // tslint:disable-next-line:no-eval
    if (offset.isInvalidConvertExpression) {
      return 'UNSUPPORTED_CONVERSION_EXPRESSION';
    }

    return new VM().run(
      replaceOffsetExpressionValue(offset, rawOffsetValue)
    );
  }
};

export const replaceOffsetExpressionValue = (offset: Offset, rawOffsetValue: RawOffsetValue): string => {
  return offset.convert.replace(
    new RegExp(/{VAL}/g),
    Array.isArray(rawOffsetValue)
      ? `${JSON.stringify(rawOffsetValue)}`
      : typeof rawOffsetValue === 'string'
        ? `'${rawOffsetValue}'`
        : rawOffsetValue.toString()
  );
};
