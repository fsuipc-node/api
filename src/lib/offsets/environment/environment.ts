import { Type } from 'fsuipc';

import { OffsetCategory } from '@shared/offset-category';
import { OffsetList } from '@shared/offset-list';
import { Offset } from '@shared/offset';

export const environment: OffsetList = {
  clockHour: new Offset({
    value: 0x238,
    name: 'clockHour',
    category: OffsetCategory.ENVIRONMENT,
    description: '0-23',
    type: Type.Byte,
    permission: 'rw',
  }),
  clockMin: new Offset({
    value: 0x239,
    name: 'clockMin',
    category: OffsetCategory.ENVIRONMENT,
    description: '0-59',
    type: Type.Byte,
    permission: 'rw',
  }),
  clockSec: new Offset({
    value: 0x23A,
    name: 'clockSec',
    category: OffsetCategory.ENVIRONMENT,
    description: '0-59',
    type: Type.Byte,
    permission: 'rw',
  }),
  zuluHour: new Offset({
    value: 0x23B,
    name: 'zuluHour',
    category: OffsetCategory.ENVIRONMENT,
    description: '0-23 UTC',
    type: Type.Byte,
    permission: 'rw',
  }),
  zuluMin: new Offset({
    value: 0x23C,
    name: 'zuluMin',
    category: OffsetCategory.ENVIRONMENT,
    description: '0-59 UTC',
    type: Type.Byte,
    permission: 'rw',
  }),
  zuluDayOfMonth: new Offset({
    value: 0x23D,
    name: 'zuluDayOfMonth',
    category: OffsetCategory.ENVIRONMENT,
    description: 'day of month UTC - FSX only - count from 1',
    type: Type.Byte,
    permission: 'r'
  }),
  dayOfYear: new Offset({
    value: 0x23E,
    name: 'dayOfYear',
    category: OffsetCategory.ENVIRONMENT,
    description: 'day of year - count from 1',
    type: Type.UInt16,
    permission: 'rw',
  }),
  zuluYear: new Offset({
    value: 0x240,
    name: 'zuluYear',
    category: OffsetCategory.ENVIRONMENT,
    description: 'year UTC',
    type: Type.UInt16,
    permission: 'rw',
  }),
  zuluMonthOfYear: new Offset({
    value: 0x242,
    name: 'zuluMonthOfYear',
    category: OffsetCategory.ENVIRONMENT,
    description: 'month of year UTC - FSX only',
    type: Type.Byte,
    permission: 'r',
  }),
  zuluDayOfWeek: new Offset({
    value: 0x243,
    name: 'zuluDayOfWeek',
    category: OffsetCategory.ENVIRONMENT,
    description: 'day of week UTC - FSX only',
    type: Type.Byte,
    permission: 'r',
  }),
  localMonthOfYear: new Offset({
    value: 0x244,
    name: 'localMonthOfYear',
    category: OffsetCategory.ENVIRONMENT,
    description: 'local month of year - FSX only',
    type: Type.Byte,
    permission: 'r',
  }),
  localDayOfMonth: new Offset({
    value: 0x245,
    name: 'localDayOfMonth',
    category: OffsetCategory.ENVIRONMENT,
    description: 'local day of month - FSX only',
    type: Type.Byte,
    permission: 'r',
  }),
  timezoneOffsetToZulu: new Offset({
    value: 0x246,
    name: 'timezoneOffsetToZulu',
    category: OffsetCategory.ENVIRONMENT,
    description: 'timezone offset minutes to zulu - determines aircraft position (+ve = behind zulu, -ve = ahead zulu)',
    type: Type.Int16,
    permission: 'rw',
  }),
  season: new Offset({
    value: 0x248,
    name: 'season',
    category: OffsetCategory.ENVIRONMENT,
    description: 'local season',
    convert: 'seasons',
    mapping: true,
    type: Type.UInt16,
    permission: 'rw',
  }),
  localYear: new Offset({
    value: 0x24A,
    name: 'localYear',
    category: OffsetCategory.ENVIRONMENT,
    description: 'local year',
    type: Type.UInt16,
    permission: 'rw',
  }),
  nearestAirportId: new Offset({
    value: 0x658,
    name: 'nearestAirportId',
    category: OffsetCategory.ENVIRONMENT,
    description: 'nearest airport',
    type: Type.ByteArray,
    length: 128,
    convert: 'nearestAirportsIds',
    mapping: true,
    permission: 'r',
  }),
  nearestAirportLatitude: new Offset({
    value: 0x65C,
    name: 'nearestAirportLatitude',
    category: OffsetCategory.ENVIRONMENT,
    description: 'nearest airport latitude',
    type: Type.Single,
    permission: 'r',
  }),
  nearestAirportLongitude: new Offset({
    value: 0x660,
    name: 'nearestAirportLongitude',
    category: OffsetCategory.ENVIRONMENT,
    description: 'nearest airport longitude',
    type: Type.Single,
    permission: 'r',
  }),
  nearestAirportAltitude: new Offset({
    value: 0x664,
    name: 'nearestAirportAltitude',
    category: OffsetCategory.ENVIRONMENT,
    description: 'nearest airport altitude - ft',
    type: Type.Single,
    permission: 'r',
  }),
  nearestAirportDistance: new Offset({
    value: 0x668,
    name: 'nearestAirportDistance',
    category: OffsetCategory.ENVIRONMENT,
    description: 'nearest airport distance - NM',
    type: Type.Single,
    permission: 'r',
  }),
  towerLatitude: new Offset({
    value: 0xD50,
    name: 'towerLatitude',
    category: OffsetCategory.ENVIRONMENT,
    description: 'tower latitude',
    convert: '{VAL} * 90 / (10001750 * 65536 * 65536)',
    type: Type.Int64,
    permission: 'r',
  }),
  towerLongitude: new Offset({
    value: 0xD58,
    name: 'towerLongitude',
    category: OffsetCategory.ENVIRONMENT,
    description: 'tower longitude',
    convert: '{VAL} * 360 / (65536 * 65536 * 65536 * 65536)',
    type: Type.Int64,
    permission: 'r',
  }),
  towerAltitude: new Offset({
    value: 0xD60,
    name: 'towerAltitude',
    category: OffsetCategory.ENVIRONMENT,
    description: 'tower altitude',
    convert: '+({VAL} * 3.28084 / (65536 * 65536)).toFixed(2)',
    type: Type.Int64,
    permission: 'r',
  }),
  nearestWeatherStationId: new Offset({
    value: 0xE80,
    name: 'nearestWeatherStationId',
    category: OffsetCategory.ENVIRONMENT,
    description: 'nearest weather station ICAO id',
    type: Type.String,
    length: 4,
    permission: 'r',
  }),
};
