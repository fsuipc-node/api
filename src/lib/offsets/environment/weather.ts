import { Type } from 'fsuipc';

import { OffsetCategory } from '@shared/offset-category';
import { OffsetList } from '@shared/offset-list';
import { Offset } from '@shared/offset';

export const weather: OffsetList = {
  metarStationAltitude: new Offset({
    value: 0x4B4,
    name: 'metarStationAltitude',
    category: OffsetCategory.WEATHER,
    description: 'METAR station altitude - ft',
    type: Type.UInt16,
    convert: '+({VAL} * 3.28084).toFixed(2)',
    permission: 'r',
  }),
  windSurfaceTurbulenceGustSpeed: new Offset({
    value: 0x4BA,
    name: 'windSurfaceTurbulenceGustSpeed',
    category: OffsetCategory.WEATHER,
    description: 'METAR station altitude - kt',
    type: Type.UInt16,
    convert: '+({VAL}).toFixed(2)',
    permission: 'r',
  }),
  metarBarometricDrift: new Offset({
    value: 0x4BC,
    name: 'metarBarometricDrift',
    category: OffsetCategory.WEATHER,
    description: 'METAR barometric drift - difference between aircraft/METAR QNH. Adding drift will give correct value for ATIS report',
    type: Type.UInt16,
    convert: '{VAL} / 16',
    permission: 'r',
  }),
  metarVisibility: new Offset({
    value: 0x4C0,
    name: 'metarVisibility',
    category: OffsetCategory.WEATHER,
    description: 'METAR visibility - sm',
    type: Type.UInt16,
    convert: '{VAL} / 100',
    permission: 'r',
  }),
  metarCloudThunderBase: new Offset({
    value: 0x4C2,
    name: 'metarCloudThunderBase',
    category: OffsetCategory.WEATHER,
    description: 'METAR visibility - ft',
    type: Type.UInt16,
    convert: '+({VAL} * 3.28084).toFixed(2)',
    permission: 'r',
  }),
  metarCloudBaseLow: new Offset({
    value: 0x4C4,
    name: 'metarCloudBaseLow',
    category: OffsetCategory.WEATHER,
    description: 'METAR cloud base low - ft',
    type: Type.UInt16,
    convert: '+({VAL} * 3.28084).toFixed(2)',
    permission: 'r',
  }),
  metarCloudBaseHigh: new Offset({
    value: 0x4C6,
    name: 'metarCloudBaseHigh',
    category: OffsetCategory.WEATHER,
    description: 'METAR cloud base high - ft',
    type: Type.UInt16,
    convert: '+({VAL} * 3.28084).toFixed(2)',
    permission: 'r',
  }),
  dewPoint: new Offset({
    value: 0x4C8,
    name: 'dewPoint',
    category: OffsetCategory.WEATHER,
    description: 'dew point - FS2000+',
    type: Type.Int16,
    convert: '{VAL} / 256',
    permission: 'r',
  }),
  precipitationRate: new Offset({
    value: 0x4C8,
    name: 'precipitationRate',
    category: OffsetCategory.WEATHER,
    description: 'precipitation rate - FS2000+ - 0 to 5',
    type: Type.Byte,
    permission: 'r',
  }),
  precipitationType: new Offset({
    value: 0x4CC,
    name: 'precipitationType',
    category: OffsetCategory.WEATHER,
    description: 'precipitation type - FS2000+ - 0 to 5',
    type: Type.Byte,
    convert: 'precipitationType',
    mapping: true,
    permission: 'r',
  }),
  cloudThunderCoverage: new Offset({
    value: 0x4CD,
    name: 'cloudThunderCoverage',
    category: OffsetCategory.WEATHER,
    description: 'cloud thunder coverage - octa',
    type: Type.Byte,
    permission: 'r',
  }),
  cloudLowCoverage: new Offset({
    value: 0x4CE,
    name: 'cloudLowCoverage',
    category: OffsetCategory.WEATHER,
    description: 'cloud low coverage - octa',
    type: Type.Byte,
    permission: 'r',
  }),
  cloudHighCoverage: new Offset({
    value: 0x4CF,
    name: 'cloudHighCoverage',
    category: OffsetCategory.WEATHER,
    description: 'cloud high coverage - octa',
    type: Type.Byte,
    permission: 'r',
  }),
  precipitationControl: new Offset({
    value: 0x4D2,
    name: 'precipitationControl',
    category: OffsetCategory.WEATHER,
    description: 'hi-byte type 0-2 - low-byte rate 0-5 - writing 0xFFFF release control',
    type: Type.UInt16,
    permission: 'w',
  }),
  dewPointControl: new Offset({
    value: 0x4D4,
    name: 'dewPointControl',
    category: OffsetCategory.WEATHER,
    description: 'degrees / 256 - 0x8000 release control',
    type: Type.UInt16,
    permission: 'w',
  }),
  surfaceWindSpeed: new Offset({
    value: 0x4D8,
    name: 'surfaceWindSpeed',
    category: OffsetCategory.WEATHER,
    description: 'surface wind speed - FS2000+ - kts',
    type: Type.UInt16,
    permission: 'r',
  }),
  surfaceWindDirection: new Offset({
    value: 0x4DA,
    name: 'surfaceWindDirection',
    category: OffsetCategory.WEATHER,
    description: 'surface wind direction - FS2000+ - deg MAG',
    type: Type.UInt16,
    convert: '+({VAL} * 360 / 65536).toFixed(2)',
    permission: 'r',
  }),
  unlimitedVisibility: new Offset({
    value: 0xBF8,
    name: 'unlimitedVisibility',
    category: OffsetCategory.WEATHER,
    description: 'unlimited visibility value - sm',
    type: Type.Int32,
    convert: '+({VAL} / 1600).toFixed(2)',
    permission: 'rw',
  }),
  cloudTypeAtAircraftAltitude: new Offset({
    value: 0xE84,
    name: 'cloudTypeAtAircraftAltitude',
    category: OffsetCategory.WEATHER,
    description: 'cloud type at aircraft atltitude - 0-10',
    type: Type.Byte,
    permission: 'r',
  }),
  cloudCoverageAtAircraftAltitude: new Offset({
    value: 0xE85,
    name: 'cloudCoverageAtAircraftAltitude',
    category: OffsetCategory.WEATHER,
    description: 'cloud coverage at aircraft atltitude - okta',
    type: Type.Byte,
    permission: 'r',
  }),
  cloudIcingAtAircraftAltitude: new Offset({
    value: 0xE86,
    name: 'cloudIcingAtAircraftAltitude',
    category: OffsetCategory.WEATHER,
    description: 'cloud icing at aircraft atltitude - 0-4',
    type: Type.UInt16,
    permission: 'r',
  }),
  cloudTurbulenceAtAircraftAltitude: new Offset({
    value: 0xE88,
    name: 'cloudTurbulenceAtAircraftAltitude',
    category: OffsetCategory.WEATHER,
    description: 'cloud turbulence at aircraft atltitude - 0-255',
    type: Type.UInt16,
    permission: 'r',
  }),
  visibility: new Offset({
    value: 0xE8A,
    name: 'visibility',
    category: OffsetCategory.WEATHER,
    description: 'current visibility - sm',
    convert: '+({VAL} / 100).toFixed(2)',
    type: Type.Int16,
    permission: 'r',
  }),
  OAT: new Offset({
    value: 0xE8C,
    name: 'OAT',
    category: OffsetCategory.WEATHER,
    description: 'current outside air temperature (OAT) - celsius',
    convert: 'Math.round({VAL} / 256)',
    type: Type.Int16,
    permission: 'r',
  }),
  interpolatedDewPoint: new Offset({
    value: 0xE8E,
    name: 'interpolatedDewPoint',
    category: OffsetCategory.WEATHER,
    description: 'dew point interpolated for aircraft altitude - celsius',
    convert: 'Math.round({VAL} / 256)',
    type: Type.Int16,
    permission: 'r',
  }),
  windSpeed: new Offset({
    value: 0xE90,
    name: 'windSpeed',
    category: OffsetCategory.WEATHER,
    description: 'wind speed at aircraft - kt',
    type: Type.UInt16,
    permission: 'r',
  }),
  windDirection: new Offset({
    value: 0xE92,
    name: 'windDirection',
    category: OffsetCategory.WEATHER,
    description: 'wind direction at aircraft - TRUE degrees',
    convert: '+({VAL} * 360 / 65536).toFixed(2)',
    type: Type.UInt16,
    permission: 'r',
  }),
};
