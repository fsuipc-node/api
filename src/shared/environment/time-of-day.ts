export enum TimeOfDay {
  DAY = 'day',
  DUSK_DAWN = 'duskDawn',
  NIGHT = 'night',
}
