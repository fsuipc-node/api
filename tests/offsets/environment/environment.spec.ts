import { environment as offsets } from '@offsets/environment/environment';
import { replaceOffsetExpressionValue } from '@convert/apply-conversion';

describe('offset - environment/environment', () => {
  const offsetsTestCases = [
    { name: 'towerLatitude', value: 2.2329449e+16, expectedResult: 46.78263288720104 },
    { name: 'towerLongitude', value: -3.6582626e+18, expectedResult: -71.39333265196446 },
    { name: 'towerAltitude', value: 392731796979, expectedResult: 300 },
  ];

  describe('offsets list', () => {
    it('should have required properties', () => {
      expect(offsets).toMatchSnapshot();
    });
  });

  offsetsTestCases.forEach(testedOffset => {
    describe(testedOffset.name, () => {
      it('should convert data properly', () => {
        const convertExpression = replaceOffsetExpressionValue(offsets[testedOffset.name], testedOffset.value);

        // tslint:disable-next-line:no-eval
        expect(eval(convertExpression)).toEqual(testedOffset.expectedResult);
      });
    });
  });
});
