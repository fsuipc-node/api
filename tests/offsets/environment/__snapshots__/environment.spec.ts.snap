// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`offset - environment/environment offsets list should have required properties 1`] = `
Object {
  "clockHour": Offset {
    "category": "environment",
    "convert": undefined,
    "description": "0-23",
    "length": undefined,
    "mapping": undefined,
    "name": "clockHour",
    "permission": "rw",
    "type": 0,
    "value": 568,
  },
  "clockMin": Offset {
    "category": "environment",
    "convert": undefined,
    "description": "0-59",
    "length": undefined,
    "mapping": undefined,
    "name": "clockMin",
    "permission": "rw",
    "type": 0,
    "value": 569,
  },
  "clockSec": Offset {
    "category": "environment",
    "convert": undefined,
    "description": "0-59",
    "length": undefined,
    "mapping": undefined,
    "name": "clockSec",
    "permission": "rw",
    "type": 0,
    "value": 570,
  },
  "dayOfYear": Offset {
    "category": "environment",
    "convert": undefined,
    "description": "day of year - count from 1",
    "length": undefined,
    "mapping": undefined,
    "name": "dayOfYear",
    "permission": "rw",
    "type": 5,
    "value": 574,
  },
  "localDayOfMonth": Offset {
    "category": "environment",
    "convert": undefined,
    "description": "local day of month - FSX only",
    "length": undefined,
    "mapping": undefined,
    "name": "localDayOfMonth",
    "permission": "r",
    "type": 0,
    "value": 581,
  },
  "localMonthOfYear": Offset {
    "category": "environment",
    "convert": undefined,
    "description": "local month of year - FSX only",
    "length": undefined,
    "mapping": undefined,
    "name": "localMonthOfYear",
    "permission": "r",
    "type": 0,
    "value": 580,
  },
  "localYear": Offset {
    "category": "environment",
    "convert": undefined,
    "description": "local year",
    "length": undefined,
    "mapping": undefined,
    "name": "localYear",
    "permission": "rw",
    "type": 5,
    "value": 586,
  },
  "nearestAirportAltitude": Offset {
    "category": "environment",
    "convert": undefined,
    "description": "nearest airport altitude - ft",
    "length": undefined,
    "mapping": undefined,
    "name": "nearestAirportAltitude",
    "permission": "r",
    "type": 9,
    "value": 1636,
  },
  "nearestAirportDistance": Offset {
    "category": "environment",
    "convert": undefined,
    "description": "nearest airport distance - NM",
    "length": undefined,
    "mapping": undefined,
    "name": "nearestAirportDistance",
    "permission": "r",
    "type": 9,
    "value": 1640,
  },
  "nearestAirportId": Offset {
    "category": "environment",
    "convert": "nearestAirportsIds",
    "description": "nearest airport",
    "length": 128,
    "mapping": true,
    "name": "nearestAirportId",
    "permission": "r",
    "type": 10,
    "value": 1624,
  },
  "nearestAirportLatitude": Offset {
    "category": "environment",
    "convert": undefined,
    "description": "nearest airport latitude",
    "length": undefined,
    "mapping": undefined,
    "name": "nearestAirportLatitude",
    "permission": "r",
    "type": 9,
    "value": 1628,
  },
  "nearestAirportLongitude": Offset {
    "category": "environment",
    "convert": undefined,
    "description": "nearest airport longitude",
    "length": undefined,
    "mapping": undefined,
    "name": "nearestAirportLongitude",
    "permission": "r",
    "type": 9,
    "value": 1632,
  },
  "nearestWeatherStationId": Offset {
    "category": "environment",
    "convert": undefined,
    "description": "nearest weather station ICAO id",
    "length": 4,
    "mapping": undefined,
    "name": "nearestWeatherStationId",
    "permission": "r",
    "type": 11,
    "value": 3712,
  },
  "season": Offset {
    "category": "environment",
    "convert": "seasons",
    "description": "local season",
    "length": undefined,
    "mapping": true,
    "name": "season",
    "permission": "rw",
    "type": 5,
    "value": 584,
  },
  "timezoneOffsetToZulu": Offset {
    "category": "environment",
    "convert": undefined,
    "description": "timezone offset minutes to zulu - determines aircraft position (+ve = behind zulu, -ve = ahead zulu)",
    "length": undefined,
    "mapping": undefined,
    "name": "timezoneOffsetToZulu",
    "permission": "rw",
    "type": 2,
    "value": 582,
  },
  "towerAltitude": Offset {
    "category": "environment",
    "convert": "+({VAL} * 3.28084 / (65536 * 65536)).toFixed(2)",
    "description": "tower altitude",
    "length": undefined,
    "mapping": undefined,
    "name": "towerAltitude",
    "permission": "r",
    "type": 4,
    "value": 3424,
  },
  "towerLatitude": Offset {
    "category": "environment",
    "convert": "{VAL} * 90 / (10001750 * 65536 * 65536)",
    "description": "tower latitude",
    "length": undefined,
    "mapping": undefined,
    "name": "towerLatitude",
    "permission": "r",
    "type": 4,
    "value": 3408,
  },
  "towerLongitude": Offset {
    "category": "environment",
    "convert": "{VAL} * 360 / (65536 * 65536 * 65536 * 65536)",
    "description": "tower longitude",
    "length": undefined,
    "mapping": undefined,
    "name": "towerLongitude",
    "permission": "r",
    "type": 4,
    "value": 3416,
  },
  "zuluDayOfMonth": Offset {
    "category": "environment",
    "convert": undefined,
    "description": "day of month UTC - FSX only - count from 1",
    "length": undefined,
    "mapping": undefined,
    "name": "zuluDayOfMonth",
    "permission": "r",
    "type": 0,
    "value": 573,
  },
  "zuluDayOfWeek": Offset {
    "category": "environment",
    "convert": undefined,
    "description": "day of week UTC - FSX only",
    "length": undefined,
    "mapping": undefined,
    "name": "zuluDayOfWeek",
    "permission": "r",
    "type": 0,
    "value": 579,
  },
  "zuluHour": Offset {
    "category": "environment",
    "convert": undefined,
    "description": "0-23 UTC",
    "length": undefined,
    "mapping": undefined,
    "name": "zuluHour",
    "permission": "rw",
    "type": 0,
    "value": 571,
  },
  "zuluMin": Offset {
    "category": "environment",
    "convert": undefined,
    "description": "0-59 UTC",
    "length": undefined,
    "mapping": undefined,
    "name": "zuluMin",
    "permission": "rw",
    "type": 0,
    "value": 572,
  },
  "zuluMonthOfYear": Offset {
    "category": "environment",
    "convert": undefined,
    "description": "month of year UTC - FSX only",
    "length": undefined,
    "mapping": undefined,
    "name": "zuluMonthOfYear",
    "permission": "r",
    "type": 0,
    "value": 578,
  },
  "zuluYear": Offset {
    "category": "environment",
    "convert": undefined,
    "description": "year UTC",
    "length": undefined,
    "mapping": undefined,
    "name": "zuluYear",
    "permission": "rw",
    "type": 5,
    "value": 576,
  },
}
`;
